torch/nn/parallel/distributed_cpu.py

```diff
@@ -0,0 +1,112 @@
+        function of the output of this module) is a distributed synchronization
+        point. Take that into account in case different processes might be
+        executing different code.
+
+    .. warning::
+        This module assumes all parameters are registered in the model by the
+        time it is created. No parameters should be added nor removed later.
+
+    .. warning::
+        This module assumes all gradients are dense.
+
+    .. warning::
+        This module doesn't work with :func:`torch.autograd.grad` (i.e. it will
+        only work if gradients are to be accumulated in ``.grad`` attributes of
+        parameters).
+
+    .. note::
+        Parameters are broadcast between nodes in the first forward. The
+        module performs an all-reduce step on gradients and assumes that they
+        will be modified by the optimizer in all processes in the same way.
+
+    .. warning::
+        Forward and backward hooks defined on :attr:`module` and its submodules
+        won't be invoked anymore, unless the hooks are initialized in the
+        :meth:`forward` method.
+
+    Args:
+        module: module to be parallelized
+
+    Example::
+
+        >>> torch.distributed.init_process_group(world_size=4, init_method='...')
+        >>> net = torch.nn.DistributedDataParallelCPU(model)
+    """
+
+    def __init__(self, module):
+        super(DistributedDataParallelCPU, self).__init__()
+        self.module = module
+        self.first_call = True
+
+        def allreduce_params():
+            if (self.needs_reduction):
+                self.needs_reduction = False
+                buckets = {}
+                for param in self.module.parameters():
+                    if param.requires_grad and param.grad is not None:
+                        tp = type(param.data)
+                        if tp not in buckets:
+                            buckets[tp] = []
+                        buckets[tp].append(param)
+
+                for tp in buckets:
+                    bucket = buckets[tp]
+                    grads = [param.grad.data for param in bucket]
+                    coalesced = _flatten_dense_tensors(grads)
+                    dist.all_reduce(coalesced)
+                    coalesced /= dist.get_world_size()
+                    for buf, synced in zip(grads, _unflatten_dense_tensors(coalesced, grads)):
+                        buf.copy_(synced)
+
+        for param in list(self.module.parameters()):
+            def allreduce_hook(*unused):
+                param._execution_engine.queue_callback(allreduce_params)
+
+            if param.requires_grad:
+                param.register_hook(allreduce_hook)
+
+    def weight_broadcast(self):
+        # broadcase the wight in the first call
+        for param in self.module.parameters():
+            dist.broadcast(param.data, 0)
+
+    def forward(self, *inputs, **kwargs):
+        if self.first_call:
+            self.weight_broadcast()
+            self.first_call = False
```

Comment:
The `first_call` semantics don't match `DistributedDataParallel`. The sync should happen in `__init__`